codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/serde/XmlParserGenerator.kt

```diff
@@ -152,10 +184,440 @@ open class XmlParserGenerator(
+        return "${shapeName}ShapeDeserializer.kt"
+    }
+    private fun Shape.shapeDeserializer(
+        ctx: ProtocolGenerator.GenerationContext,
+        block: (fnName: String, writer: KotlinWriter) -> Unit,
+    ): Symbol {
+        val target = targetOrSelf(ctx.model)
+        val shapeName = StringUtils.capitalize(target.id.getName(ctx.service))
+        val symbol = ctx.symbolProvider.toSymbol(this)
+
+        val fnName = "deserialize${shapeName}Shape"
+        return buildSymbol {
+            name = fnName
+            namespace = ctx.settings.pkg.serde
+            definitionFile = shapeDeserializerDefinitionFile(ctx)
+            reference(symbol, SymbolReference.ContextOption.DECLARE)
+            renderBy = {
+                block(fnName, it)
+            }
+        }
+    }
+
+    private fun deserializeShape(
+        ctx: ProtocolGenerator.GenerationContext,
+        shape: Shape,
+        block: KotlinWriter.() -> Unit,
+    ): Symbol {
+        val symbol = ctx.symbolProvider.toSymbol(shape)
+        val deserializeFn = shape.shapeDeserializer(ctx) { fnName, writer ->
+            writer.withBlock(
+                "internal fun #L(reader: #T): #T {",
+                "}",
+                fnName,
+                SerdeXml.XmlTagReader,
+                symbol,
+            ) {
+                block(this)
+            }
+        }
+        return deserializeFn
+    }
+
+    private fun deserializeList(
+        ctx: ProtocolGenerator.GenerationContext,
+        serdeCtx: SerdeCtx,
+        member: MemberShape,
+        writer: KotlinWriter,
+    ) {
+        val target = ctx.model.expectShape<CollectionShape>(member.target)
+        val targetMember = target.member
+        val isSparse = target.hasTrait<SparseTrait>()
+        val deserializeFn = deserializeShape(ctx, target) {
+            write("val result = mutableListOf<#T#L>()", ctx.symbolProvider.toSymbol(targetMember), nullabilitySuffix(isSparse))
+            deserializeLoop(SerdeCtx(tagReader = "reader")) { innerCtx ->
+                val memberName = targetMember.getTrait<XmlNameTrait>()?.value ?: targetMember.memberName
+                withBlock("#S -> {", "}", memberName) {
+                    deserializeListInner(ctx, innerCtx, target, this)
+                    write("result.add(el)")
+                }
+            }
+            write("return result")
+        }
+        writer.write("#T(${serdeCtx.tagReader})", deserializeFn)
+    }
+
+    private fun flatCollectionAccumulatorExpr(
+        ctx: ProtocolGenerator.GenerationContext,
+        member: MemberShape,
+    ): String =
+        when (val container = ctx.model.expectShape(member.container)) {
+            is StructureShape -> "builder.${member.defaultName()}"
+            is UnionShape -> {
+                val unionVariantName = member.unionVariantName()
+                "value?.as${unionVariantName}OrNull()"
+            }
```

Comment:
It may not have been set yet. This is just the expression to use for the accumulator when dealing with flat maps/lists. We pull out the existing member from the type (structure/union) and append to the existing collection if it exists otherwise we create a new one.